internal/store/store.go

```diff
@@ -1,53 +1,91 @@
+		updateMappingDeployment("", deploy, namespace, identifier)
 		updateRoute(identifier, Route{
-			Service:   service,
-			Port:      port,
-			Label:     label,
-			Namespace: namespace,
-			LastUsed:  time.Now(), // default to time.Now()
+			Service:    service,
+			Port:       port,
+			Deployment: deploy,
+			Namespace:  namespace,
+			Domains:    domains,
+			LastUsed:   time.Now(), // default to time.Now()
 		})
 	} else { // update route
-		// TODO should handle the domain removed from the service or change in the label
+		updateDomains(route.Domains, domains, identifier)
+		updateMappingDeployment(route.Deployment, deploy, namespace, identifier)
 		route.Service = service
 		route.Port = port
-		route.Label = label
+		route.Deployment = deploy
 		route.Namespace = namespace
-		updateRoute(identifier, route)
+		route.Domains = domains
+		updateRoute(identifier, *route)
 	}
-	for _, domain := range domains {
-		updateMapping(domain, identifier)
+}
+
+func updateDomains(oldDomains, newDomains []string, identifier string) {
+	// get domains that are not in the new list
+	var domainsToBeRemoved []string
+	for _, domain := range oldDomains {
+		if !contains(newDomains, domain) {
+			domainsToBeRemoved = append(domainsToBeRemoved, domain)
+		}
+	}
+
+	// remove the domains that are not in the new list
+	for _, domain := range domainsToBeRemoved {
+		deleteMappingByKey(domain)
+	}
+
+	for _, domain := range newDomains {
+		updateMappingDomain(domain, identifier)
 	}
-	updateMapping(label, identifier)
+}
+
+func updateMappingDomain(domain, identifier string) {
+	updateMapping(domain, identifier)
+}
+
+func updateMappingDeployment(oldName, newName, namespace, identifier string) {
+	if oldName != newName { // need to remove the old name from the map
+		deleteMappingByKey(genDeploymentKey(oldName, namespace))
+	}
+	updateMapping(genDeploymentKey(newName, namespace), identifier)
 }
 
 func DeleteObjectInStore(identifier string) {
 	deleteRoute(identifier)
 	deleteMappingByValue(identifier)
 }
 
-func GetRouteByDomain(key string) (Route, error) {
-	return getRouteByMapping(key)
+func GetRouteByDomainKey(domain string) (*Route, error) {
+	return getRouteByMapping(domain)
 }
 
-func GetRouteByLabel(key string) (Route, error) {
-	return getRouteByMapping(key)
+func GetRouteByDeploymentKey(name, namespace string) (*Route, error) {
+	return getRouteByMapping(genDeploymentKey(name, namespace))
 }
 
-func getRouteByMapping(key string) (Route, error) {
-	if identifier, err := getMapping(key); err != nil {
-		return Route{}, err
+func getRouteByMapping(key string) (*Route, error) {
+	if identifier, err := getMappingValueByKey(key); err != nil {
+		return nil, err
 	} else {
 		return getRoute(identifier)
 	}
 }
 
-func UpdateLastUse(key string) {
-	if identifier, err := getMapping(key); err == nil {
+func UpdateLastUse(domain string) {
+	if identifier, err := getMappingValueByKey(domain); err == nil {
 		updateLastUse(identifier)
```

Comment:
Comment - If you imagine having many use cases for updating a route and don't want to create a unique function per use case, you can use the visitors pattern.

```
updateRoute("key", func(route) {
   route.Name = "test"
}

func updateRoute(key string, f func(Route)) {
  lock.Lock()
  defer lock.Unlock()
  r, _ = map[key] // skipping check here just for the example
  f(r)
}
```

Reply:
Wow fuck. I did not know about that.
That is awesome!
I always wondered if u could provide a function as a parameter!